selfdrive/loggerd/v4l_encoder.cc

```diff
@@ -0,0 +1,380 @@
+    .count = count
+  };
+  checked_ioctl(fd, VIDIOC_REQBUFS, &reqbuf);
+}
+
+// TODO: writing should be moved to loggerd
+void V4LEncoder::write_handler(V4LEncoder *e, const char *path) {
+  VideoWriter writer(path, e->filename, e->remuxing, e->width, e->height, e->fps, !e->remuxing, false);
+
+  bool exit = false;
+  while (!exit) {
+    auto out_buf = e->to_write.pop();
+
+    capnp::FlatArrayMessageReader cmsg(*out_buf);
+    cereal::Event::Reader event = cmsg.getRoot<cereal::Event>();
+
+    auto edata = (e->type == DriverCam) ? event.getDriverEncodeData() :
+      ((e->type == WideRoadCam) ? event.getWideRoadEncodeData() :
+      (e->remuxing ? event.getQRoadEncodeData() : event.getRoadEncodeData()));
+    auto flags = edata.getFlags();
+    if (flags & V4L2_QCOM_BUF_FLAG_EOS) exit = true;
+
+    // dangerous cast from const, but should be fine
+    auto data = edata.getData();
+    if (data.size() > 0) {
+      writer.write((uint8_t *)data.begin(), data.size(), edata.getTimestampEof(),
+        flags & V4L2_QCOM_BUF_FLAG_CODECCONFIG, flags & V4L2_BUF_FLAG_KEYFRAME);
+    }
+
+    // free the data
+    delete out_buf;
+  }
+
+  // VideoWriter is freed on out of scope
+}
+
+void V4LEncoder::dequeue_handler(V4LEncoder *e) {
+  e->segment_num++;
+  uint32_t idx = 0;
+  bool exit = false;
+
+  // POLLIN is capture, POLLOUT is frame
+  struct pollfd pfd;
+  pfd.events = POLLIN | POLLOUT;
+  pfd.fd = e->fd;
+
+  while (!exit) {
+    int rc = poll(&pfd, 1, 1000);
+    if (!rc) { printf("poll timeout"); continue; }
+
+    /*struct timespec t;
+    clock_gettime(CLOCK_BOOTTIME, &t);
+    uint64_t current_time = t.tv_sec * 1000000000ULL + t.tv_nsec;
+    printf("poll %x at %f ms\n", pfd.revents, current_time/1e6);*/
+
+    if (pfd.revents & POLLIN) {
+      unsigned int bytesused, flags, index;
+      struct timeval timestamp;
+      int ret = dequeue_buffer(e->fd, V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE, &index, &bytesused, &flags, &timestamp);
+      assert(ret == 0);
+      e->buf_out[index].sync(VISIONBUF_SYNC_FROM_DEVICE);
+      uint8_t *buf = (uint8_t*)e->buf_out[index].addr;
+
+      // eof packet, we exit
+      if (flags & V4L2_QCOM_BUF_FLAG_EOS) exit = true;
+      uint64_t ts = timestamp.tv_sec * 1000000 + timestamp.tv_usec;
+
+      // broadcast packet
+      MessageBuilder msg;
+      auto event = msg.initEvent(true);
+      auto edata = (e->type == DriverCam) ? event.initDriverEncodeData() :
+        ((e->type == WideRoadCam) ? event.initWideRoadEncodeData() :
+        (e->remuxing ? event.initQRoadEncodeData() : event.initRoadEncodeData()));
+      edata.setData(kj::arrayPtr<capnp::byte>(buf, bytesused));
+      edata.setTimestampEof(ts);
+      edata.setIdx(idx++);
+      edata.setSegmentNum(e->segment_num);
+      edata.setFlags(flags);
+
+      auto words = new kj::Array<capnp::word>(capnp::messageToFlatArray(msg));
+      auto bytes = words->asBytes();
+      e->pm->send(e->service_name, bytes.begin(), bytes.size());
```

Comment:
Any reason you're not using the overloaded version?

```suggestion
      e->pm->send(e->service_name, msg);
```

Reply:
Because I need the Array and not an ArrayPtr so it doesn't free on out of scope.